File: switcher-library/src/main/java/pl/aprilapps/switcher/Switcher.java
package pl.aprilapps.switcher;

import android.content.Context;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * Created by Jacek Kwiecień on 14.03.15.
 */
public class Switcher {

    private List<View> contentViews = new ArrayList<>();
    private List<View> errorViews = new ArrayList<>();
    private List<View> emptyViews = new ArrayList<>();
    private List<View> progressViews = new ArrayList<>();

    private TextView errorLabel;
    private TextView progressLabel;
    private int animDuration = 300;

    private List<Animations.FadeListener> runningAnimators = new ArrayList<>();

    private Switcher() {

    }

    private void addContentView(View contentView) {
        this.contentViews.add(contentView);
    }

    private void addErrorView(View errorView) {
        this.errorViews.add(errorView);
    }

    private void addEmptyView(View emptyView) {
        this.emptyViews.add(emptyView);
    }


    private void addProgressView(View progressView) {
        this.progressViews.add(progressView);
    }

    private void setErrorLabel(TextView errorLabel) {
        this.errorLabel = errorLabel;
    }

    private void setProgressLabel(TextView progressLabel) {
        this.progressLabel = progressLabel;
    }

    private void setAnimationDuration(int duration) {
        animDuration = duration;
    }

    public static class Builder {

        private Switcher switcher = new Switcher();
        private Context context;

        public Builder(Context context) {
            this.context = context;
            switcher.setAnimationDuration(context.getResources().getInteger(android.R.integer.config_shortAnimTime));
        }

        public Builder addContentView(View contentView) {
            switcher.addContentView(contentView);
            return this;
        }

        public Builder addErrorView(View errorView) {
            switcher.addErrorView(errorView);
            return this;
        }

        public Builder addEmptyView(View emptyView) {
            switcher.addEmptyView(emptyView);
            return this;
        }

        public Builder addProgressView(View progressView) {
            switcher.addProgressView(progressView);
            return this;
        }

        public Builder setProgressLabel(TextView progressLabel) {
            switcher.setProgressLabel(progressLabel);
            return this;
        }

        public Builder setErrorLabel(TextView errorLabel) {
            switcher.setErrorLabel(errorLabel);
            return this;
        }

        public Switcher build() {
            switcher.setupViews();
            return switcher;
        }
    }

    private void setupViews() {
        for (View contentView : contentViews) {
            contentView.setVisibility(View.VISIBLE);
        }

        for (View errorView : errorViews) {
            errorView.setVisibility(View.INVISIBLE);
        }

        for (View errorView : errorViews) {
            errorView.setVisibility(View.INVISIBLE);
        }

        for (View progressView : progressViews) {
            progressView.setVisibility(View.INVISIBLE);
        }

        for (View emptyView : emptyViews) {
            emptyView.setVisibility(View.INVISIBLE);
        }
    }

    private List<View> membersExcludingGroup(List<View> excludedGroup) {
        List<View> members = new ArrayList<>();
        members.addAll(contentViews);
        members.addAll(errorViews);
        members.addAll(progressViews);
        members.addAll(emptyViews);

        if (excludedGroup != null && excludedGroup.size() > 0) {
            members.removeAll(excludedGroup);
        }

        return members;
    }

    private boolean isViewPartOfSwitcher(View potentialMember) {
        List<View> members = new ArrayList<>();
        members.addAll(contentViews);
        members.addAll(errorViews);
        members.addAll(progressViews);
        members.addAll(emptyViews);

        for (View view : members) {
            if (view.equals(potentialMember)) return true;
        }
        return false;
    }

    private void showGroup(List<View> groupToShow, boolean immediately) {
        cancelCurrentAnimators();

        for (View view : groupToShow) {
            if (immediately) {
                view.setVisibility(View.VISIBLE);
            } else {
                Animations.FadeListener animator = Animations.fadeIn(view, animDuration);
                runningAnimators.add(animator);
            }
        }

        List<View> viewsToHide = membersExcludingGroup(contentViews);
        for (View view : viewsToHide) {
            if (immediately) {
                view.setVisibility(View.INVISIBLE);
            } else {
                Animations.FadeListener animator = Animations.fadeOut(view, animDuration);
                runningAnimators.add(animator);
            }
        }
    }

    public void showContentView() {
        Log.i(Switcher.class.getSimpleName(), "showContentView");
        showGroup(contentViews, false);
    }

    public void showContentViewImmediately() {
        Log.i(Switcher.class.getSimpleName(), "showContentViewImmediately");
        showGroup(contentViews, true);
    }


    public void showProgressView() {
        Log.i(Switcher.class.getSimpleName(), "showProgressView");
        showGroup(progressViews, false);
    }

    public void showProgressViewImmediately() {
        Log.i(Switcher.class.getSimpleName(), "showProgressViewImmediately");
        showGroup(progressViews, true);
    }

    public void showErrorView() {
        Log.i(Switcher.class.getSimpleName(), "showErrorView");
        showGroup(errorViews, false);
    }

    public void showErrorViewImmediately() {
        Log.i(Switcher.class.getSimpleName(), "showErrorView");
        showGroup(errorViews, true);
    }

    public void showEmptyView() {
        Log.i(Switcher.class.getSimpleName(), "showEmptyView");
        showGroup(emptyViews, false);
    }

    public void showEmptyViewImmediately() {
        Log.i(Switcher.class.getSimpleName(), "showEmptyView");
        showGroup(emptyViews, true);
    }

    private void cancelCurrentAnimators() {
        Iterator<Animations.FadeListener> it = runningAnimators.iterator();
        while (it.hasNext()) {
            Animations.FadeListener animator = it.next();
            animator.onAnimationEnd();
            it.remove();
        }
    }

    public void showErrorView(String errorMessage) {
        if (errorLabel == null) {
            throw new NullPointerException("You have to build Switcher using withErrorLabel() method");
        }

        errorLabel.setText(errorMessage);
        showErrorView();
    }

    public void showErrorView(final OnErrorViewListener listener) {
        for (View errorView : errorViews) {
            errorView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    listener.onErrorViewClicked();
                    view.setOnClickListener(null);
                }
            });
        }
        showErrorView();
    }

    public void showErrorView(String errorMessage, final OnErrorViewListener listener) {
        if (errorLabel == null) {
            throw new NullPointerException("You have to build Switcher using withErrorLabel() method");
        }

        errorLabel.setText(errorMessage);
        showErrorView(listener);
    }

    public void showProgressView(String errorMessage) {
        if (errorLabel == null) {
            throw new NullPointerException("You have to build Switcher using withProgressLabel() method");
        }

        progressLabel.setText(errorMessage);
        showProgressView();
    }

}

File: library/src/main/java/pl/aprilapps/switcher/Switcher.kt
package pl.aprilapps.switcher

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.content.Context
import android.util.Log
import android.view.View
import android.widget.TextView

/**
 * Created by Jacek Kwiecień on 04.03.2017.
 */
open class Switcher(context: Context) {

    private val contentViews = mutableSetOf<View>()
    private val progressViews = mutableSetOf<View>()
    private val errorViews = mutableSetOf<View>()
    private val emptyViews = mutableSetOf<View>()
    private var progressLabel: TextView? = null
    private var errorLabel: TextView? = null
    private var invisibleState = View.INVISIBLE
    private var logsEnabled = false

    var animationDuration = context.resources.getInteger(android.R.integer.config_shortAnimTime).toLong()

    var animatorSet = AnimatorSet()
    val runningAnimators = mutableListOf<Animator>()

    private fun areAllContentViewsVisible(): Boolean {
        contentViews.forEach { if (it.visibility != View.VISIBLE) return false }
        return true
    }

    private fun areAllProgressViewsVisible(): Boolean {
        progressViews.forEach { if (it.visibility != View.VISIBLE) return false }
        return true
    }

    private fun areAllErrorViewsVisible(): Boolean {
        errorViews.forEach { if (it.visibility != View.VISIBLE) return false }
        return true
    }

    private fun areAllEmptyViewsVisible(): Boolean {
        emptyViews.forEach { if (it.visibility != View.VISIBLE) return false }
        return true
    }

    fun startAnimationsPackage(animators: List<Animator>) {
        printLog("ENDING currently running animations: " + runningAnimators.size)
        runningAnimators.forEach { it.end() }
        runningAnimators.clear()
        animatorSet = AnimatorSet()
        animatorSet.playTogether(animators)
        printLog("STARTING animations package: " + animators.size + " animators.")
        animatorSet.start()
    }

    private fun printLog(message: String) {
        if (logsEnabled) Log.v(javaClass.simpleName, message)
    }

    private fun printWarningLog(message: String) {
        if (logsEnabled) Log.w(javaClass.simpleName, message)
    }


    private fun printErrorLog(message: String) {
        if (logsEnabled) Log.e(javaClass.simpleName, message)
    }

    private fun fadeInAnimator(view: View, animationDuration: Long): Animator {
        val animator = ObjectAnimator.ofFloat(view, "alpha", 0f, 1f).setDuration(animationDuration)
        animator.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationEnd(animator: Animator) {
                animator.removeAllListeners()
                val iterator = runningAnimators.iterator()
                iterator.takeIf { it == animator }?.forEach { iterator.remove() }
                printLog("fade IN animation ENDED: " + animator.toString())
            }

            override fun onAnimationStart(animator: Animator) {
                printLog("fade IN animation STARTED: " + animator.toString())
                view.visibility = View.VISIBLE
            }
        })

        runningAnimators.add(animator)
        return animator
    }

    private fun fadeOutAnimator(view: View, animationDuration: Long): Animator {
        val animator = ObjectAnimator.ofFloat(view, "alpha", 1f, 0f).setDuration(animationDuration)
        animator.addListener(object : AnimatorListenerAdapter() {

            override fun onAnimationEnd(animator: Animator) {
                animator.removeAllListeners()
                val iterator = runningAnimators.iterator()
                iterator.takeIf { it == animator }?.forEach { iterator.remove() }
                printLog("fade OUT animation ENDED: " + animator.toString())
                view.visibility = invisibleState
            }

            override fun onAnimationStart(animator: Animator) {
                printLog("fade OUT animation STARTED: " + animator.toString())
            }
        })

        runningAnimators.add(animator)
        return animator
    }

    //    ===================   CONTENT   ===================

    private fun showContentView(duration: Long) {
        if (areAllContentViewsVisible()) {
            printWarningLog("All content views are currently visible. Aborting.")
            return
        }

        val animators = mutableListOf<Animator>()
                .plus(contentViews.map { fadeInAnimator(it, duration) })
                .plus(allNonContentViews().filter { it.visibility != invisibleState }.map { fadeOutAnimator(it, duration) })
        startAnimationsPackage(animators)
    }

    fun showContentView() {
        printLog("Showing content")
        showContentView(animationDuration)
    }

    fun showContentViewImmediately() {
        printLog("Showing content immediately")
        showContentView(0)
    }

    //    ===================   PROGRESS   ===================

    private fun showProgressView(duration: Long, progressMessage: String?) {
        progressMessage?.let { progressLabel?.let { progressLabel!!.text = progressMessage } }

        if (areAllProgressViewsVisible()) {
            printWarningLog("All progress views are currently visible. Aborting.")
            return
        }

        val animators = mutableListOf<Animator>()
                .plus(progressViews.map { fadeInAnimator(it, duration) })
                .plus(allNonProgressViews().filter { it.visibility != invisibleState }.map { fadeOutAnimator(it, duration) })
        startAnimationsPackage(animators)
    }

    fun showProgressView() {
        showProgressView(animationDuration, null)
    }

    fun showProgressViewImmediately() {
        printLog("Showing progress immediately")
        showProgressView(0, null)
    }

    fun showProgressView(progressMessage: String?) {
        printLog("Showing progress")
        showProgressView(animationDuration, progressMessage)
    }

    fun showProgressViewImmediately(progressMessage: String?) {
        printLog("Showing progress immediately")
        showProgressView(0, progressMessage)
    }


    //    ===================   ERROR   ===================

    private fun showErrorView(duration: Long, errorMessage: String?) {
        errorMessage?.let { errorLabel?.let { errorLabel!!.text = errorMessage } }

        if (areAllErrorViewsVisible()) {
            printWarningLog("All error views are currently visible. Aborting.")
            return
        }

        val animators = mutableListOf<Animator>()
                .plus(errorViews.map { fadeInAnimator(it, duration) })
                .plus(allNonErrorViews().filter { it.visibility != invisibleState }.map { fadeOutAnimator(it, duration) })
        startAnimationsPackage(animators)
    }

    fun showErrorView() {
        printLog("Showing error")
        showErrorView(animationDuration, null)
    }

    fun showErrorViewImmediately() {
        printLog("Showing error immediately")
        showErrorView(0, null)
    }

    fun showErrorView(errorMessage: String?) {
        printLog("Showing error")
        showErrorView(animationDuration, errorMessage)
    }

    fun showErrorViewImmediately(errorMessage: String?) {
        printLog("Showing error immediately")
        showErrorView(0, errorMessage)
    }

    //    ===================   ERROR   ===================

    private fun showEmptyView(duration: Long) {
        if (areAllEmptyViewsVisible()) {
            printWarningLog("All empty views are currently visible. Aborting.")
            return
        }

        val animators = mutableListOf<Animator>()
        animators.addAll(emptyViews.map { fadeInAnimator(it, duration) })
        animators.addAll(allNonEmptyViews().map { fadeOutAnimator(it, duration) })
        startAnimationsPackage(animators)
    }

    fun showEmptyView() {
        printLog("Showing empty view")
        showEmptyView(animationDuration)
    }

    fun showEmptyViewImmediately() {
        printLog("Showing empty view immediately")
        showEmptyView(0)
    }

    private fun allNonContentViews(): MutableSet<View> {
        val set = mutableSetOf<View>()
        set.addAll(progressViews)
        set.addAll(errorViews)
        set.addAll(emptyViews)
        return set
    }

    private fun allNonErrorViews(): MutableSet<View> {
        val set = mutableSetOf<View>()
        set.addAll(contentViews)
        set.addAll(progressViews)
        set.addAll(emptyViews)
        return set
    }

    private fun allNonEmptyViews(): MutableSet<View> {
        val set = mutableSetOf<View>()
        set.addAll(contentViews)
        set.addAll(progressViews)
        set.addAll(errorViews)
        return set
    }

    private fun allNonProgressViews(): MutableSet<View> {
        val set = mutableSetOf<View>()
        set.addAll(contentViews)
        set.addAll(errorViews)
        set.addAll(emptyViews)
        return set
    }

    private fun setupViewsInitialVisibility() {
        printLog("=======================================")
        printLog("Building Switcher. Content views: " + contentViews.size + ", error views: " + errorViews.size + ", progress views: " + progressViews.size + ", empty views: " + emptyViews.size)
        contentViews.plus(progressViews).plus(errorViews).plus(emptyViews).forEach { it.visibility = View.VISIBLE }
        allNonContentViews().forEach { it.visibility = invisibleState }
        contentViews.forEach { it.visibility = View.VISIBLE }
    }

    open class Builder(context: Context) {
        val switcher = Switcher(context)

        fun setAnimDuration(millis: Long): Builder {
            switcher.animationDuration = millis
            return this
        }

        fun setInvisibleStateToGone() {
            switcher.invisibleState = View.GONE
        }

        fun setErrorLabel(errorLabel: TextView): Builder {
            switcher.errorLabel = errorLabel
            return this
        }

        fun setProgressLabel(progressLabel: TextView): Builder {
            switcher.progressLabel = progressLabel
            return this
        }

        fun addContentView(view: View): Builder {
            switcher.contentViews.add(view)
            return this
        }

        fun addProgressView(view: View): Builder {
            switcher.progressViews.add(view)
            return this
        }

        fun addErrorView(view: View): Builder {
            switcher.errorViews.add(view)
            return this
        }

        fun addEmptyView(view: View): Builder {
            switcher.emptyViews.add(view)
            return this
        }

        fun setLogsEnabled(enabled: Boolean): Builder {
            switcher.logsEnabled = enabled
            return this
        }

        fun build(): Switcher {
            switcher.setupViewsInitialVisibility()
            return switcher
        }
    }

}